compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -211,8 +211,14 @@ object Inliner {
     val targs = fun match
       case TypeApply(_, targs) => targs
       case _ => Nil
+
     val unapplyInfo = sym.info match
-      case info: PolyType => info.instantiate(targs.map(_.tpe))
+      case info: PolyType => info.instantiate(targs.map(_.tpe)) match
+        case MethodTpe(_, _, rt: PolyType) => rt.instantiate(targs.map(_.tpe))
```

Comment:
For the cases `MethodTpe(_, _, rt: PolyType)` we assume that these are extension methods because no other definition have that type. I would make this a bit more explicit and add the flag test for this branch as well.
```diff
-        case MethodTpe(_, _, rt: PolyType) => rt.instantiate(targs2.map(_.tpe))
+        case MethodTpe(_, _, rt: PolyType) if sym.flags.is(ExtensionMethod) =>
+          rt.instantiate(targs2.map(_.tpe))
```

Reply:
We could also refactor the common code and make the logic related to extensions clearer.

```scala
     def extensionInfo(info: Type, targs: List[Tree]): Type =
       info match
        case MethodTpe(_, _, rt: PolyType) => rt.instantiate(targs.map(_.tpe))
        case MethodTpe(_, _, rt) => rt
        case info => info
     val unapplyInfo = sym.info match
      case pt: PolyType =>
        val info = pt.instantiate(targs1.map(_.tpe))
        if sym.flags.is(ExtensionMethod) then extensionInfo(info, targs2) else info
      case info =>
        if sym.flags.is(ExtensionMethod) then extensionInfo(info, targs1) else info
```

Reply:
Widening might not be the correct thing to do. We might be missing some test cases. I imagine that if we have path dependent types widening will make us loose precision.

Reply:
One that comes to my mind but is not supported due to some other bug is https://github.com/lampepfl/dotty/pull/13158#issuecomment-890853536. We could imagine a similar unapply with a path dependent type on a term that is not a parameter. Also, we might try to have a path dependent type in the trailing given parameters of the unapply.

Reply:
I'm not sure about the example you give in the comment but the code example in #12991 compiles and runs correctly in this branch